lib/controllers/esri-token-auth.js

```diff
@@ -0,0 +1,86 @@
+/* jshint node: true, esnext: true */
+"use strict";
+var router = require('express').Router();
+var request = require('request');
+var bodyParser = require('body-parser');
+var url = require('url');
+
+module.exports = function(options) {
+    if (!options || !options.servers) {
+        return;
+    }
+
+    // The maximum size of the JSON data.
+    let postSizeLimit = options.postSizeLimit || '1024';
+
+    let tokenServers = parseUrls(options.servers);
+
+    router.use(bodyParser.json({limit:postSizeLimit, type:'application/json'}));
+    router.post('/', function(req, res, next) {
+        let parameters = req.body;
+
+        if (!parameters.url || parameters.url.length === 0) {
+            return res.status(400).send('No URL specified.');
+        }
+
+        let targetUrl = parseURL(parameters.url);
+        if (!targetUrl || targetUrl.length === 0 && (typeof targetUrl === string)) {
+            return res.status(400).send('Invalid URL specified.');
+        }
+
+        let tokenServer = tokenServers[targetUrl];
+        if (!tokenServer) {
+            return res.status(400).send('Unsupported URL specified.');
+        }
+
+        if (!tokenServer.username || !tokenServer.password || !tokenServer.tokenUrl) {
+            console.error("Bad Configuration. " + targetUrl + " does not supply all of the required properties.");
+            return res.status(400).send('Invalid server configuration.');
+        }
+
+        request({
+            url: tokenServer.tokenUrl,
+            method: 'POST',
+            headers: {
+                'User-Agent': 'TerriaJSESRITokenAuth',
+            },
+            form:{
+                username: tokenServer.username,
+                password: tokenServer.password,
+                f: 'JSON'
+            }
+        }, function(error, response, body) {
+            res.set('Content-Type', 'application/json');
+
+            if (response.statusCode != 200) {
+                return res.status(400).send('Token server failed.');
+            } else {
+                let value = JSON.parse(response.body);
+                return res.status(200).send(JSON.stringify(value));
+            }
+        });
+    });
+
+    router.use(function(err, req, res, next) {
```

Comment:
What's the purpose of this?

Reply:
Not sure whether I understand this question correctly but it is to handle JSON parser errors as well as any other possible errors with the router. Since the JSON is coming from the client it can be completely invalid:
https://expressjs.com/en/guide/error-handling.html

Have added a comment to try and clarify this. 

Can generate an appropriate error using:
`curl -X POST --data "[object]" localhost:3001/esri-token-auth --header "Content-Type:application/json"`

4e24c4757fef983862645350de6773caf2cc7e76

Reply:
Ok, I don't think it makes sense to handle all errors by returning a generic 400 and "Unknown error" though.  Specific errors (e.g. user provided bad data) should be handled by returning an appropriate 4xx code directly in the main handler.  "Unknown" errors can just be handled by the default express behavior of returning a 500 (internal server error), because an unknown error is the server's fault not the user's.

Reply:
I'm not exactly sure what you mean here, since I think that this is the only way to get a handle on these errors (they don't seem to propagate through `router.post` since they are generated from the `router.use(bodyParser.json(`).

Take the point on status codes, but still not sure on whether there is a clean way of differentiating them.

Reply:
doesn't `bodyParser.json()` automatically return a 4xx code then?

Reply:
Its a `<title>500 - Internal server error</title>`, but I guess this is cleaner.

Have removed.

913312d209ce5d1d5f72d2b6f66ab2baffa02472

Reply:
500?  that's really frustrating.  If the body is supposed to be JSON and it's not, that's pretty clearly an error on the part of the caller, not an internal server error.  I'd call that a bug in bodyParser.  Anyway, I'm not that concerned about it.

Reply:
>  I'd call that a bug in bodyParser.

Yeah agreed.